
%% bare_jrnl.tex
%% V1.3
%% 2007/01/11
%% by Michael Shell
%% see http://www.michaelshell.org/
%% for current contact information.
%%
%% This is a skeleton file demonstrating the use of IEEEtran.cls
%% (requires IEEEtran.cls version 1.7 or later) with an IEEE journal paper.
%%
%% Support sites:
%% http://www.michaelshell.org/tex/ieeetran/
%% http://www.ctan.org/tex-archive/macros/latex/contrib/IEEEtran/
%% and
%% http://www.ieee.org/



% *** Authors should verify (and, if needed, correct) their LaTeX system  ***
% *** with the testflow diagnostic prior to trusting their LaTeX platform ***
% *** with production work. IEEE's font choices can trigger bugs that do  ***
% *** not appear when using other class files.                            ***
% The testflow support page is at:
% http://www.michaelshell.org/tex/testflow/


%%*************************************************************************
%% Legal Notice:
%% This code is offered as-is without any warranty either expressed or
%% implied; without even the implied warranty of MERCHANTABILITY or
%% FITNESS FOR A PARTICULAR PURPOSE! 
%% User assumes all risk.
%% In no event shall IEEE or any contributor to this code be liable for
%% any damages or losses, including, but not limited to, incidental,
%% consequential, or any other damages, resulting from the use or misuse
%% of any information contained here.
%%
%% All comments are the opinions of their respective authors and are not
%% necessarily endorsed by the IEEE.
%%
%% This work is distributed under the LaTeX Project Public License (LPPL)
%% ( http://www.latex-project.org/ ) version 1.3, and may be freely used,
%% distributed and modified. A copy of the LPPL, version 1.3, is included
%% in the base LaTeX documentation of all distributions of LaTeX released
%% 2003/12/01 or later.
%% Retain all contribution notices and credits.
%% ** Modified files should be clearly indicated as such, including  **
%% ** renaming them and changing author support contact information. **
%%
%% File list of work: IEEEtran.cls, IEEEtran_HOWTO.pdf, bare_adv.tex,
%%                    bare_conf.tex, bare_jrnl.tex, bare_jrnl_compsoc.tex
%%*************************************************************************

% Note that the a4paper option is mainly intended so that authors in
% countries using A4 can easily print to A4 and see how their papers will
% look in print - the typesetting of the document will not typically be
% affected with changes in paper size (but the bottom and side margins will).
% Use the testflow package mentioned above to verify correct handling of
% both paper sizes by the user's LaTeX system.
%
% Also note that the "draftcls" or "draftclsnofoot", not "draft", option
% should be used if it is desired that the figures are to be displayed in
% draft mode.
%
\documentclass[journal]{IEEEtran}
\usepackage{blindtext}
\usepackage{graphicx}
\usepackage{float}
\usepackage{csquotes}
\usepackage{caption}
% Some very useful LaTeX packages include:
% (uncomment the ones you want to load)


% *** MISC UTILITY PACKAGES ***
%
%\usepackage{ifpdf}
% Heiko Oberdiek's ifpdf.sty is very useful if you need conditional
% compilation based on whether the output is pdf or dvi.
% usage:
% \ifpdf
%   % pdf code
% \else
%   % dvi code
% \fi
% The latest version of ifpdf.sty can be obtained from:
% http://www.ctan.org/tex-archive/macros/latex/contrib/oberdiek/
% Also, note that IEEEtran.cls V1.7 and later provides a builtin
% \ifCLASSINFOpdf conditional that works the same way.
% When switching from latex to pdflatex and vice-versa, the compiler may
% have to be run twice to clear warning/error messages.






% *** CITATION PACKAGES ***
%
%\usepackage{cite}
% cite.sty was written by Donald Arseneau
% V1.6 and later of IEEEtran pre-defines the format of the cite.sty package
% \cite{} output to follow that of IEEE. Loading the cite package will
% result in citation numbers being automatically sorted and properly
% "compressed/ranged". e.g., [1], [9], [2], [7], [5], [6] without using
% cite.sty will become [1], [2], [5]--[7], [9] using cite.sty. cite.sty's
% \cite will automatically add leading space, if needed. Use cite.sty's
% noadjust option (cite.sty V3.8 and later) if you want to turn this off.
% cite.sty is already installed on most LaTeX systems. Be sure and use
% version 4.0 (2003-05-27) and later if using hyperref.sty. cite.sty does
% not currently provide for hyperlinked citations.
% The latest version can be obtained at:
% http://www.ctan.org/tex-archive/macros/latex/contrib/cite/
% The documentation is contained in the cite.sty file itself.






% *** GRAPHICS RELATED PACKAGES ***
%
\ifCLASSINFOpdf
  % \usepackage[pdftex]{graphicx}
  % declare the path(s) where your graphic files are
  % \graphicspath{{../pdf/}{../jpeg/}}
  % and their extensions so you won't have to specify these with
  % every instance of \includegraphics
  % \DeclareGraphicsExtensions{.pdf,.jpeg,.png}
\else
  % or other class option (dvipsone, dvipdf, if not using dvips). graphicx
  % will default to the driver specified in the system graphics.cfg if no
  % driver is specified.
  % \usepackage[dvips]{graphicx}
  % declare the path(s) where your graphic files are
  % \graphicspath{{../eps/}}
  % and their extensions so you won't have to specify these with
  % every instance of \includegraphics
  % \DeclareGraphicsExtensions{.eps}
\fi
% graphicx was written by David Carlisle and Sebastian Rahtz. It is
% required if you want graphics, photos, etc. graphicx.sty is already
% installed on most LaTeX systems. The latest version and documentation can
% be obtained at: 
% http://www.ctan.org/tex-archive/macros/latex/required/graphics/
% Another good source of documentation is "Using Imported Graphics in
% LaTeX2e" by Keith Reckdahl which can be found as epslatex.ps or
% epslatex.pdf at: http://www.ctan.org/tex-archive/info/
%
% latex, and pdflatex in dvi mode, support graphics in encapsulated
% postscript (.eps) format. pdflatex in pdf mode supports graphics
% in .pdf, .jpeg, .png and .mps (metapost) formats. Users should ensure
% that all non-photo figures use a vector format (.eps, .pdf, .mps) and
% not a bitmapped formats (.jpeg, .png). IEEE frowns on bitmapped formats
% which can result in "jaggedy"/blurry rendering of lines and letters as
% well as large increases in file sizes.
%
% You can find documentation about the pdfTeX application at:
% http://www.tug.org/applications/pdftex


%URL package for url links in the bibliography
\usepackage{url}


% *** MATH PACKAGES ***
%
\usepackage[cmex10]{amsmath}
\usepackage{bm}
\usepackage{booktabs}
\usepackage{listings}

% A popular package from the American Mathematical Society that provides
% many useful and powerful commands for dealing with mathematics. If using
% it, be sure to load this package with the cmex10 option to ensure that
% only type 1 fonts will utilized at all point sizes. Without this option,
% it is possible that some math symbols, particularly those within
% footnotes, will be rendered in bitmap form which will result in a
% document that can not be IEEE Xplore compliant!
%
% Also, note that the amsmath package sets \interdisplaylinepenalty to 10000
% thus preventing page breaks from occurring within multiline equations. Use:
\interdisplaylinepenalty=2500
% after loading amsmath to restore such page breaks as IEEEtran.cls normally
% does. amsmath.sty is already installed on most LaTeX systems. The latest
% version and documentation can be obtained at:
% http://www.ctan.org/tex-archive/macros/latex/required/amslatex/math/

\usepackage{color}

\definecolor{mygreen}{rgb}{0,0.6,0}
\definecolor{mygray}{rgb}{0.5,0.5,0.5}
\definecolor{mymauve}{rgb}{0.58,0,0.82}

\lstset{ %
  backgroundcolor=\color{white},   % choose the background color; you must add \usepackage{color} or \usepackage{xcolor}
  basicstyle=\footnotesize,        % the size of the fonts that are used for the code
  breakatwhitespace=false,         % sets if automatic breaks should only happen at whitespace
  breaklines=true,                 % sets automatic line breaking
  captionpos=b,                    % sets the caption-position to bottom
  commentstyle=\color{mygreen},    % comment style
  deletekeywords={...},            % if you want to delete keywords from the given language
  escapeinside={\%*}{*)},          % if you want to add LaTeX within your code
  extendedchars=true,              % lets you use non-ASCII characters; for 8-bits encodings only, does not work with UTF-8
  frame=single,	                   % adds a frame around the code
  keepspaces=true,                 % keeps spaces in text, useful for keeping indentation of code (possibly needs columns=flexible)
  keywordstyle=\color{blue},       % keyword style
  language=Octave,                 % the language of the code
  otherkeywords={*,...},           % if you want to add more keywords to the set
  numbers=left,                    % where to put the line-numbers; possible values are (none, left, right)
  numbersep=5pt,                   % how far the line-numbers are from the code
  numberstyle=\tiny\color{mygray}, % the style that is used for the line-numbers
  rulecolor=\color{black},         % if not set, the frame-color may be changed on line-breaks within not-black text (e.g. comments (green here))
  showspaces=false,                % show spaces everywhere adding particular underscores; it overrides 'showstringspaces'
  showstringspaces=false,          % underline spaces within strings only
  showtabs=false,                  % show tabs within strings adding particular underscores
  stepnumber=2,                    % the step between two line-numbers. If it's 1, each line will be numbered
  stringstyle=\color{mymauve},     % string literal style
  tabsize=2,	                   % sets default tabsize to 2 spaces
  title=\lstname                   % show the filename of files included with \lstinputlisting; also try caption instead of title
}





% *** SPECIALIZED LIST PACKAGES ***
%
%\usepackage{algorithmic}
% algorithmic.sty was written by Peter Williams and Rogerio Brito.
% This package provides an algorithmic environment fo describing algorithms.
% You can use the algorithmic environment in-text or within a figure
% environment to provide for a floating algorithm. Do NOT use the algorithm
% floating environment provided by algorithm.sty (by the same authors) or
% algorithm2e.sty (by Christophe Fiorio) as IEEE does not use dedicated
% algorithm float types and packages that provide these will not provide
% correct IEEE style captions. The latest version and documentation of
% algorithmic.sty can be obtained at:
% http://www.ctan.org/tex-archive/macros/latex/contrib/algorithms/
% There is also a support site at:
% http://algorithms.berlios.de/index.html
% Also of interest may be the (relatively newer and more customizable)
% algorithmicx.sty package by Szasz Janos:
% http://www.ctan.org/tex-archive/macros/latex/contrib/algorithmicx/




% *** ALIGNMENT PACKAGES ***
%
%\usepackage{array}
% Frank Mittelbach's and David Carlisle's array.sty patches and improves
% the standard LaTeX2e array and tabular environments to provide better
% appearance and additional user controls. As the default LaTeX2e table
% generation code is lacking to the point of almost being broken with
% respect to the quality of the end results, all users are strongly
% advised to use an enhanced (at the very least that provided by array.sty)
% set of table tools. array.sty is already installed on most systems. The
% latest version and documentation can be obtained at:
% http://www.ctan.org/tex-archive/macros/latex/required/tools/


%\usepackage{mdwmath}
%\usepackage{mdwtab}
% Also highly recommended is Mark Wooding's extremely powerful MDW tools,
% especially mdwmath.sty and mdwtab.sty which are used to format equations
% and tables, respectively. The MDWtools set is already installed on most
% LaTeX systems. The lastest version and documentation is available at:
% http://www.ctan.org/tex-archive/macros/latex/contrib/mdwtools/


% IEEEtran contains the IEEEeqnarray family of commands that can be used to
% generate multiline equations as well as matrices, tables, etc., of high
% quality.


%\usepackage{eqparbox}
% Also of notable interest is Scott Pakin's eqparbox package for creating
% (automatically sized) equal width boxes - aka "natural width parboxes".
% Available at:
% http://www.ctan.org/tex-archive/macros/latex/contrib/eqparbox/





% *** SUBFIGURE PACKAGES ***
%\usepackage[tight,footnotesize]{subfigure}
% subfigure.sty was written by Steven Douglas Cochran. This package makes it
% easy to put subfigures in your figures. e.g., "Figure 1a and 1b". For IEEE
% work, it is a good idea to load it with the tight package option to reduce
% the amount of white space around the subfigures. subfigure.sty is already
% installed on most LaTeX systems. The latest version and documentation can
% be obtained at:
% http://www.ctan.org/tex-archive/obsolete/macros/latex/contrib/subfigure/
% subfigure.sty has been superceeded by subfig.sty.



%\usepackage[caption=false]{caption}
%\usepackage[font=footnotesize]{subfig}
% subfig.sty, also written by Steven Douglas Cochran, is the modern
% replacement for subfigure.sty. However, subfig.sty requires and
% automatically loads Axel Sommerfeldt's caption.sty which will override
% IEEEtran.cls handling of captions and this will result in nonIEEE style
% figure/table captions. To prevent this problem, be sure and preload
% caption.sty with its "caption=false" package option. This is will preserve
% IEEEtran.cls handing of captions. Version 1.3 (2005/06/28) and later 
% (recommended due to many improvements over 1.2) of subfig.sty supports
% the caption=false option directly:
%\usepackage[caption=false,font=footnotesize]{subfig}
%
% The latest version and documentation can be obtained at:
% http://www.ctan.org/tex-archive/macros/latex/contrib/subfig/
% The latest version and documentation of caption.sty can be obtained at:
% http://www.ctan.org/tex-archive/macros/latex/contrib/caption/




% *** FLOAT PACKAGES ***
%
%\usepackage{fixltx2e}
% fixltx2e, the successor to the earlier fix2col.sty, was written by
% Frank Mittelbach and David Carlisle. This package corrects a few problems
% in the LaTeX2e kernel, the most notable of which is that in current
% LaTeX2e releases, the ordering of single and double column floats is not
% guaranteed to be preserved. Thus, an unpatched LaTeX2e can allow a
% single column figure to be placed prior to an earlier double column
% figure. The latest version and documentation can be found at:
% http://www.ctan.org/tex-archive/macros/latex/base/



%\usepackage{stfloats}
% stfloats.sty was written by Sigitas Tolusis. This package gives LaTeX2e
% the ability to do double column floats at the bottom of the page as well
% as the top. (e.g., "\begin{figure*}[!b]" is not normally possible in
% LaTeX2e). It also provides a command:
%\fnbelowfloat
% to enable the placement of footnotes below bottom floats (the standard
% LaTeX2e kernel puts them above bottom floats). This is an invasive package
% which rewrites many portions of the LaTeX2e float routines. It may not work
% with other packages that modify the LaTeX2e float routines. The latest
% version and documentation can be obtained at:
% http://www.ctan.org/tex-archive/macros/latex/contrib/sttools/
% Documentation is contained in the stfloats.sty comments as well as in the
% presfull.pdf file. Do not use the stfloats baselinefloat ability as IEEE
% does not allow \baselineskip to stretch. Authors submitting work to the
% IEEE should note that IEEE rarely uses double column equations and
% that authors should try to avoid such use. Do not be tempted to use the
% cuted.sty or midfloat.sty packages (also by Sigitas Tolusis) as IEEE does
% not format its papers in such ways.


%\ifCLASSOPTIONcaptionsoff
%  \usepackage[nomarkers]{endfloat}
% \let\MYoriglatexcaption\caption
% \renewcommand{\caption}[2][\relax]{\MYoriglatexcaption[#2]{#2}}
%\fi
% endfloat.sty was written by James Darrell McCauley and Jeff Goldberg.
% This package may be useful when used in conjunction with IEEEtran.cls'
% captionsoff option. Some IEEE journals/societies require that submissions
% have lists of figures/tables at the end of the paper and that
% figures/tables without any captions are placed on a page by themselves at
% the end of the document. If needed, the draftcls IEEEtran class option or
% \CLASSINPUTbaselinestretch interface can be used to increase the line
% spacing as well. Be sure and use the nomarkers option of endfloat to
% prevent endfloat from "marking" where the figures would have been placed
% in the text. The two hack lines of code above are a slight modification of
% that suggested by in the endfloat docs (section 8.3.1) to ensure that
% the full captions always appear in the list of figures/tables - even if
% the user used the short optional argument of \caption[]{}.
% IEEE papers do not typically make use of \caption[]'s optional argument,
% so this should not be an issue. A similar trick can be used to disable
% captions of packages such as subfig.sty that lack options to turn off
% the subcaptions:
% For subfig.sty:
% \let\MYorigsubfloat\subfloat
% \renewcommand{\subfloat}[2][\relax]{\MYorigsubfloat[]{#2}}
% For subfigure.sty:
% \let\MYorigsubfigure\subfigure
% \renewcommand{\subfigure}[2][\relax]{\MYorigsubfigure[]{#2}}
% However, the above trick will not work if both optional arguments of
% the \subfloat/subfig command are used. Furthermore, there needs to be a
% description of each subfigure *somewhere* and endfloat does not add
% subfigure captions to its list of figures. Thus, the best approach is to
% avoid the use of subfigure captions (many IEEE journals avoid them anyway)
% and instead reference/explain all the subfigures within the main caption.
% The latest version of endfloat.sty and its documentation can obtained at:
% http://www.ctan.org/tex-archive/macros/latex/contrib/endfloat/
%
% The IEEEtran \ifCLASSOPTIONcaptionsoff conditional can also be used
% later in the document, say, to conditionally put the References on a 
% page by themselves.





% *** PDF, URL AND HYPERLINK PACKAGES ***
%
%\usepackage{url}
% url.sty was written by Donald Arseneau. It provides better support for
% handling and breaking URLs. url.sty is already installed on most LaTeX
% systems. The latest version can be obtained at:
% http://www.ctan.org/tex-archive/macros/latex/contrib/misc/
% Read the url.sty source comments for usage information. Basically,
% \url{my_url_here}.





% *** Do not adjust lengths that control margins, column widths, etc. ***
% *** Do not use packages that alter fonts (such as pslatex).         ***
% There should be no need to do such things with IEEEtran.cls V1.6 and later.
% (Unless specifically asked to do so by the journal or conference you plan
% to submit to, of course. )


% correct bad hyphenation here
\hyphenation{op-tical net-works semi-conduc-tor}


\begin{document}
%
% paper title
% can use linebreaks \\ within to get better formatting as desired
\title{Solving Travelling Salesman Problems with Genetic Algorithms}
%
%
% author names and IEEE memberships
% note positions of commas and nonbreaking spaces ( ~ ) LaTeX will not break
% a structure at a ~ so this keeps an author's name from being broken across
% two lines.
% use \thanks{} to gain access to the first footnote area
% a separate \thanks must be used for each paragraph as LaTeX2e's \thanks
% was not built to handle multiple paragraphs
%

\author{Samuel Jackson, University of Aberystwyth}

% note the % following the last \IEEEmembership and also \thanks - 
% these prevent an unwanted space from occurring between the last author name
% and the end of the author line. i.e., if you had this:
% 
% \author{....lastname \thanks{...} \thanks{...} }
%                     ^------------^------------^----Do not want these spaces!
%
% a space would be appended to the last name and could cause every name on that
% line to be shifted left slightly. This is one of those "LaTeX things". For
% instance, "\textbf{A} \textbf{B}" will typeset as "A B" not "AB". To get
% "AB" then you have to do: "\textbf{A}\textbf{B}"
% \thanks is no different in this regard, so shield the last } of each \thanks
% that ends a line with a % and do not let a space in before the next \thanks.
% Spaces after \IEEEmembership other than the last one are OK (and needed) as
% you are supposed to have spaces between the names. For what it is worth,
% this is a minor point as most people would not even notice if the said evil
% space somehow managed to creep in.



% The paper headers
%\markboth{Journal of \LaTeX\ Class Files,~Vol.~6, No.~1, January~2007}%
%{Shell \MakeLowercase{\textit{et al.}}: Bare Demo of IEEEtran.cls for Journals}
% The only time the second header will appear is for the odd numbered pages
% after the title page when using the twoside option.
% 
% *** Note that you probably will NOT want to include the author's ***
% *** name in the headers of peer review papers.                   ***
% You can use \ifCLASSOPTIONpeerreview for conditional compilation here if
% you desire.




% If you want to put a publisher's ID mark on the page you can do it like
% this:
%\IEEEpubid{0000--0000/00\$00.00~\copyright~2007 IEEE}
% Remember, if you use this you must call \IEEEpubidadjcol in the second
% column for its text to clear the IEEEpubid mark.



% use for special paper notices
%\IEEEspecialpapernotice{(Invited Paper)}




% make the title area
\maketitle


%\begin{abstract}
%\boldmath
%\blindtext[1]
%\end{abstract}
% IEEEtran.cls defaults to using nonbold math in the Abstract.
% This preserves the distinction between vectors and scalars. However,
% if the journal you are submitting to favors bold math in the abstract,
% then you can use LaTeX's standard command \boldmath at the very start
% of the abstract to achieve this. Many IEEE journals frown on math
% in the abstract anyway.

% Note that keywords are not normally used for peerreview papers.
%\begin{IEEEkeywords}
%IEEEtran, journal, \LaTeX, paper, template.
%\end{IEEEkeywords}






% For peer review papers, you can put extra information on the cover
% page as needed:
% \ifCLASSOPTIONpeerreview
% \begin{center} \bfseries EDICS Category: 3-BBND \end{center}
% \fi
%
% For peerreview papers, this IEEEtran command inserts a page break and
% creates the second title. It will be ignored for other modes.
\IEEEpeerreviewmaketitle



\section{Introduction}

\noindent 
A genetic algorithm (GA) is an optimisation method frequently used to find approximate solutions in challenging problem domains. Genetic algorithms are inspired by the biological concept of natural selection. This places genetic algorithms under the category of biologically inspired approaches to optimisation; along with genetic programming, ant colony optimisation, and particle swarm optimisation. Many traditional optimisation techniques rely on the calculation of derivatives and often require detailed knowledge of the search space. GAs on the other hand only require a measure of solution quality, making them well suited to difficult optimisation problems where traditional techniques would otherwise fail.
In a GA solutions are encoded as “chromosomes”. A chromosome is usually an array of binary, integer, or real numbers but other representations are possible. For example, an array of real numbers might represent the encoding of coefficients of a polynomial in a curve fitting problem. A ``gene'' in GA terminology is a single atomic component of a chromosome. In the previous curve fitting example a gene would be a single coefficient.

A GA proceeds by creating an initial population of randomly generated solutions. From this population a subset of candidates are selected and used to generate the next population. New solutions are generated from this subset using crossover and selection operators. Crossover creates new chromosomes from two or more parent chromosomes by combining portions from each of the parent chromosomes together. Crossover aims to preserve some information about what makes decent solutions between generations. Mutation randomly modifies a chromosome by altering one or more of its genes. The mutation operator aims to encourage exploration of the search space and avoid local minima. Finally, each chromosome in the new population is evaluated according a measure of quality known as the ``fitness''. The fitness function is the entirely dependant on the problem domain. In the curve fitting example above the fitness function could be the mean squared error of the polynomial represented by a particular chromosome.

\begin{figure}[H]
\centering
\includegraphics[width=0.5\textwidth]{img/tsp_solution_example.png}
\caption{Example of a solution to a TSP with 30 cities distributed uniformly at random. Each of the blue points represents a city. The red lines indicate the order of traversal between cities.}
\label{fig:tsp-example}
\end{figure}

The travelling salesman problem (TSP) is a classic mathematical problem. The TSP is defined as follows:

\begin{displayquote}
\textit{The traveling salesman problem is a problem ... requiring the most efficient (i.e., least total distance) Hamiltonian cycle a salesman can take through each of $n$ cities.}\cite{weisstein2006traveling}
\end{displayquote}
The problem is simple to understand but computationally intensive to calculate an exact solution. The TSP has been shown to be NP-hard \cite{weisstein2006traveling} and a brute force solution requires $O(n!)$ time.  A GA makes no guarantees about finding an optimal solution to a TSP, but can be used to find an approximate solution in a reasonable amount of time given decent parameters. This is faster than a basic brute force search because a GA will only examine a subset of solutions in the search space that may or may not include the optimum result, but should converge towards the optimum solution with good parameters.

\section{Program Description}
The program for this assignment is implemented as a Python package with a basic command line interface (CLI). The installation of the package and operation of the CLI are described in detail in appendices \ref{appendix:installation} and \ref{appendix:cli}.

The main implementation of the GA algorithm is within the sub-package \textit{tspsolver.ga}. This contains a separate module for each of the components of the GA. It also includes the \textit{simulator} module, which is responsible for composing each of the components and running the GA itself.

Each of the modules contains an abstract base class and several subclasses providing concrete implementations of types of the component. For example, the crossover module contains a class \textit{AbstractCrossoverOperator} which implements operations common to all crossover operators and provides an abstract method \textit{\_crossover\_for\_chromosomes} which all subclasses must implement. This allows the \textit{Simulator} class to use each component without knowledge of the implementation details of the concrete component. In software engineering terminology this is known as the ``strategy'' design pattern. The strategy pattern enables the behaviour of the algorithm to be dynamically selected at runtime. 

\begin{figure}[H]
\centering
\includegraphics[width=0.5\textwidth]{img/class_diagram.png}
\caption{Class diagram for the simulator class and its components using the strategy pattern. The simulator is composed with one of each type of component. The abstract classes provide a common interface for each of the components. Concrete implementations of the components derive from the relevant base class}
\label{fig:class-diagram}
\end{figure}


This implementation takes the pattern slightly further. The \textit{Simulator} class also dynamically loads the correct components using Python's reflection capabilities according to user supplied parameters. The parameters for the application are supplied via the CLI in the form of a JSON file. Information regarding the structure of the parameter file can be found in the README and information on the parameters in appendix \ref{appendix:parameters}.

Once instantiated the \textit{Simulator} takes a 2D matrix with two columns ($x$ and $y$ coordinates) and $n$ rows (equal to the number of cities) as a parameters to the \textit{evolve} method. This method converts the datasets into a distance matrix which is used to approximate a solution using the GA.

In addition to the main \textit{tspsolver.ga} sub-package there are four other modules. These contain supporting code for running the simulator and setting up the CLI. These modules are:

\begin{itemize}
	\item \textbf{tsp\_generator} - Implements a class for generating TSP datasets with uniformly random cities.
	\item \textbf{command} - Implements the CLI and includes a couple of file I/O routines.
	\item \textbf{plotting} - Defines a couple of custom plotting functions.
	\item \textbf{tuning} - Defines a class for running a grid search over a range of GA parameters.
\end{itemize}

\section{Representation Discussion}
The experiments in this report  use a path based representation. In a path based representation each chromosome is an array of positive integers. The value of each integer represents a single city. In this program an integer corresponds to the $i^{th}$ row in the dataset. Each integer's position in the array indicates the order in which it will be visited. For example: in $[3, 4, 1, 5, 2]$ the first city to be visited would be city $3$, followed by city $4$ and then city $1$ etc. 

Since each city in the tour must only be visited once any valid solution only contains a single occurrence of each city. Therefore all valid solutions must be of size $n$ where $n$ is the number of cities. In this representation valid solutions are therefore simply permutations of the enumeration of every city in the dataset.

While many different representations of the TSP are possible \cite{larranaga1999genetic}, a path based representation remains the most natural. It is both intuitive to understand and has many different genetic operators which can be utilised. Other approaches are possible as discussed in ref. \cite{larranaga1999genetic}, but these are shown to either have poor results or have implementation difficulties. For example, a binary representation has some seemingly valid encodings which represent no valid city.

However, this does not mean that a path based representation is without limitations. The primary issues with encoding a set of cities as a chromosome for a TSP is ensuring that the genetic operators used will produce valid tours.  It is easy to see how traditional naive genetic operators would create invalid tours by producing a solution that visits the same city twice.

\section{Algorithm Components \& Genetic Operators Discussion}
There are three main types of genetic operators that are typically used as part of a GA: selection, crossover, and mutation. This application implements a variety of different genetic operators. Additionally two different techniques for generating the initial populations are provided.

\subsection{Population Generation}
The first type of population generation technique is purely random. The \textit{SimplePopulationGenerator} in the \textit{population\_generation} module creates the desired population size by creating random permutations of the list of cities. 

Random initialisation is a sensible, general baseline. However, it is obvious that some initial solutions will be better than others. In an optimal solution each city will have a city with a short distance from it next in the sequence. In general, a good heuristic may be to use the closest neighbours to a city as the adjacent entires in the chromosome. 

This will not always yield great solutions. Contemplate the closest neighbours to a few points in \ref{fig:tsp-example} and you will see that the $k^{th}$ nearest element is not necessarily the $k^{th}$ element in the tour. However, especially for larger datasets, the immediate neighbours of one particular city are more likely to end up closer together.

With this intuition a second population generator \textit{KNNPopulationGenerator} \cite{pullan2003adapting} is implemented which uses the $K$-nearest neighbours (KNN) of a city to generate a chromosome. The algorithm works as follows: for $i^{th}$ member of the population it picks the $i^{th}$ city. An ordered list of the KNN's to that city are found where $K$ is equal to the $n$. The aim is that this is more likely to lead to initial chromosomes which have some part of their chromosome already close to the correct place compared with choosing at random. To prevent a dramatic reduction in diversity, only a portion of the population is initilised in this way.

\subsection{Selection}
In GAs, selection is how individuals from the current population are chosen to produce a new population. A good selection technique should yield more ``good'' chromosomes for reproduction than bad ones. This is because the best solutions in the current population are more likely to produce even better offspring in the following generations.

This doesn't mean that the best solutions should always be chosen because this rapidly leads to a lack of diversity. Consider a candidate TSP solution that yields a poor fitness because just one of the connections geographically cuts across the whole ``world''. In this scenario the solution is ranked as poor, but in reality it is quite close to being optimal! A selection approach that naively selects just the best looking solutions is more likely to head towards local optima and would potentially fail to explore the full search space.

The experiments in this report only utilise a single type of selection operator. Originally, two types of selection operator were going to be compared. These were \textit{roulette wheel selection} (RWS) and \textit{tournament selection} (TS) \cite{colin2002genetic}. However RWS was dropped due to slow performance and poor convergence on larger datasets in comparison with TS. RWS is well known to suffer from stochastic noise leading to a large variation in the expected distribution.

In TS a small subset of the total population is chosen at random and compared against each other. This program implements \textit{strict} TS,  where the chromosome with the best fitness always ``wins'' the tournament. An alternative formulation would be to use \textit{soft} tournaments where the winner is probabilistically selected proportionally to their fitness.

\textit{Tournament} selection is a very practical technique that is simple to implement, fast, and can be parallelised. Another key advantage over RWS is that TS provides a parameter that adjusts the selection pressure applied by the  operator. Selection pressure is the likelihood that an average individual will be chosen over the fittest individual. Changing the tournament size influences how likely it is that weaker chromosomes will survive the selection process. Bigger tournaments lead to an increased chance of a better individual entering the tournament causing them to be selected and visa versa. However, it is worth noting that TS, like RWS, still suffers from stochastic noise but the effects are dampened thanks to selection pressure.

\subsection{Crossover}
The crossover operator is used to recombine selected individuals to form a new population that is distinct from (and hopefully better) than the previous population. A good crossover operator should preserve the best portions of a chromosome. Without the counter effect of mutation a good crossover operator should naturally cause the GA to converge to a (possibly optimal) solution.

The crossover operators implemented in this application can all be found in the \textit{crossover} module. There are three distinct operators, two of which are fairly similar. The first two operators are \textit{OnePointPMX} and \textit{TwoPointPMX}. Both of these are variants of \textit{partially mapped crossover} (PMX). As their names suggest the only different in their implementation is the number of pivot points used to produce sub-tours.

PMX crossover begins by copying a sub-tour of one parent chromosome to the child. This can lead to an invalid tour because there will be duplicate cities in the resulting solution. PMX repairs the new chromosomes using the mapping between the elements replaced when copying the sub-tour. The algorithm proceeds by iterating over the parts of the chromosome outside of the copied sub-tour. If a duplicate element is found then it is replaced by inserting the corresponding element in the parent chromosome using the mapping. 

PMX is both conceptually simple to understand and to implement. It is similar to regular one and two point crossover but produces valid TSP tours. This makes it a good base line to compare other techniques against.

\begin{figure}[H]
\centering
\includegraphics[width=0.4\textwidth]{img/pmx_diagram.jpg}
\caption{Shows how the \textit{TwoPointPMX} operator creates new valid children. First, sub-tours from the parent are copied to the children. Then the chromosomes are repaired by looking at what was replaced in the original parent.}
\label{fig:pmx-diagram}
\end{figure}

The other crossover operator is \textit{OrderCrossover} (OX) \cite{moscato1989genetic}. Order crossover begins in a similar manner to PMX. Sub-tours from both chromosomes are chosen from the parents at random and copied to the children. The algorithm differs in how it repairs the chromosomes. To repair the chromosome each element starting from the second pivot point is copied from the second parent to the child in order, skipping those which are already in the copied sub tour. The OX operator has an advantage over PMX because it attempts to preserve the order of the chromosomes that were not copied from the original tour. This should help to prevent the crossover accidentally being too destructive.

\begin{figure}[H]
\centering
\includegraphics[width=0.5\textwidth]{img/order_diagram.jpg}
\caption{Shows how the OX operator creates new valid children. First, sub-tours from the parent are copied to the child. Then the chromosome is repaired using by copying elements from the second parent in the order they appear, excluding ones which are already included.}
\label{fig:order-diagram}
\end{figure}

\begin{table*}[t]
\centering
\input{tables/cross_vs_mutate.tex}
\caption{Median fitness of running each of the different types of crossover and mutation with each other over $5$ randomly generated datasets each containing $50$ points. All parameter sets had a crossover rate equal to $0.9$ and a mutation rate of $0.1$. Tournament selection was used with population size of $20$ and tournament size of $5$. Each was run for a total of $1000$ generations.}
\label{table:cross-vs-mutate}
\end{table*}

\begin{table*}[t]
\centering
\input{tables/cross_vs_mutate_params.tex}
\caption{Median fitness of running each of the different types of crossover and mutation parameters with \textit{OrderCrossover} and \textit{InversionMutation}. Tournament selection was used with population size of $20$ and tournament size of $5$. Each was run for a total of $1000$ generations.}
\label{table:cross-vs-mutate-params}
\end{table*}


\subsection{Mutation}
The mutation operator is used to randomly modify chromosomes after crossover. Mutation operators, in contrast to crossover, aim to prevent the algorithm from converging too early and attempt to diversify the population.

This program implements four types of mutation operator suitable for TSP problems \cite{larranaga1999genetic}. They are: \textit{SwapCity}, \textit{Displacement}, \textit{Inversion} and \textit{Insertion} mutations.

The \textit{SwapCityMutation} chooses two cities at random and swaps their positions. This is perhaps the simplest mutation operator and therefore makes a good base line to compare other operators against.

The \textit{DisplacementMutation} is the next logical step up from \textit{SwapCityMutation}. In this operator a whole sub-tour is moved within the chromosome. The motivation behind this is that swapping individual cities is often too destructive. Moving a sub-tour should result in being less destructive as good sub-tours are preserved while still exploring the search space.

The \textit{InversionMutation} takes \textit{DisplacementMutation} another step further. \textit{InversionMutation} works identically to  \textit{DisplacementMutation} but the sub-tour is inserted in the reverse order. In addition to the benefits of \textit{DisplacementMutation} this can help solve cases where the GA has found a good route, but is hampered by a few long connections at the end of a tour. Reversing the sub-tour before insertion can solve this by removing some of the long connections through chance.

Finally, in \textit{InsertionMutation} a single gene is removed from a chromosome and reinserted in another place. This is likely to be useful in cases where a single city is linked to other cities much further away from itself. By randomly inserting the city in a different place it may be that the distance travelled to and from it is reduced.

\section{Experiments Performed}
\label{sec:experiments}
This section describes the experiments performed using the system described in the preceding sections. For all experiments, unless otherwise noted, the following setup is used: A parameter grid is generated for each combination of parameters. Each set of parameters is tested on a total of $5$ randomly generated datasets each with $50$ cities. This number was chosen to be high enough to show the differences between parameters, but low enough to make computation time realistic. All parameter sets in the test use the same datasets to ensure consistency. The median fitness value is taken to represent the whole parameter set. This ensures that the outcome of the test was not just random chance or affected by a single outlier. 

\subsection{Crossover and Mutation operators}
\label{subsec:cross-vs-mutate-operators}
The first experiment examines how crossover and mutation operators work together. Table \ref{table:cross-vs-mutate} shows the results comparing the crossover operators against mutation operators. Figure \ref{fig:cross-vs-mutate} shows the fitness of each operator combination over $1000$ epochs for a single trial.

\begin{figure}[H]
\centering
\includegraphics[width=0.5\textwidth]{figures/cross_vs_mutate_convergence.png}
\caption{Convergence of each of the crossover and mutation operator pairs on a sample dataset. Each line is the minimum (best) fitness achieved for every 10 generations. It can be seen that Order crossover with an Displacement and Inversion mutation are the quickest to make progress initially, while in this instance Order crossover with Insertion mutation finds the best solution}
\label{fig:cross-vs-mutate}
\end{figure}

\subsection{Crossover and Mutation Parameters}
\label{subsec:crossover-mutation-parameters}
Motivated by the results of this experiment and constrained by the size limits of this report and the time complexity of calculating many runs on multiple parameters sets the findings for the remaining experiments will use \textit{OrderCrossover} with \textit{InversionMutation}.

Table \ref{table:cross-vs-mutate-params} shows the results of running a grid search over a small range of possible values for the probability of crossover and mutation. The general trend is towards a larger crossover and mutation rate.

\begin{table*}[t]
\centering
\input{tables/selection_vs_pop_size.tex}
\caption{Median fitness of running each of the different parameter values for population size and tournament size with \textit{TournamentSelection}. For all tests \textit{OrderCrossover} and \textit{InversionMutation} are used. All runs used a $0.9$ crossover rate and a $0.2$ mutation rate based on the results in section \ref{subsec:crossover-mutation-parameters}. Each was run for a total of $1000$ generations.}
\label{table:selection-vs-pop-size}
\end{table*}

\begin{figure}[H]
\centering
\includegraphics[width=0.5\textwidth]{figures/cross_vs_mutate_params_convergence.png}
\caption{Convergence of the GA over a range of different parameters for crossover and mutation. All trials used \textit{OrderCrossover} and \textit{InversionMutation}. Each line is the minimum (best) fitness achieved for every 10 generations. Selection was performed using tournament selection with tournament size of $5$ and a population size of $20$. The number of points used for each trial was $50$.}
\label{fig:cross-vs-mutate-params}
\end{figure}


\begin{figure}[H]
\centering
\includegraphics[width=0.5\textwidth]{figures/selection_vs_pop_size_convergence.png}
\caption{Convergence of each of the different tournament sizes over a range of different population sizes for a single trial run. Each line is the minimum (best) fitness achieved for every 10 generations. Generally larger population and tournament sizes produce quicker convergence.}
\label{fig:tournament-selection-convergence}
\end{figure}


\begin{table*}[t]
\centering
\input{tables/knn_fitness.tex}
\caption{The fitness for a batch of $5$ datasets each containing $100$ cities with different population initialisation strategies.}
\label{table:knn-fitness}
\end{table*}

\newpage

\begin{figure}[H]
\centering
\includegraphics[width=0.5\textwidth]{figures/knn_convergence.png}
\caption{Shows the convergence of the different settings for population generation for one trial over a dataset of $100$ cities. It can be clearly seen that the KNN approach leads to faster convergence.}
\label{fig:knn-convergence}
\end{figure}

\subsection{Selection Operators}
\label{subsec:selection-parameters}
This experiment examines the selection parameters \textit{TournamentSelection}. Once again this has been limited to just using \textit{OrderCrossover} and \textit{InversionMutation}. Crossover and mutation rates are fixed to $0.9$ and $0.2$ based on the results of the previous section. Table \ref{table:selection-vs-pop-size} shows the results for the trials on population size versus tournament size for \textit{TournamentSelection}. Figure \ref{fig:tournament-selection-convergence} shows a single trial run with varying population and tournament sizes.


\subsection{Population Initialisation}
Table \ref{table:knn-fitness} shows the effect of running the same experimental setup with different population generators. For the \textit{KNNPopulationGenerator} a parameter specifies the proportion of the chromosomes that are generated by taking the KNN neighbours. Figure \ref{fig:knn-convergence} shows the convergence of one trial for each parameter set. In both cases datasets containing $100$ cities were used. This is to better accentuate the differences.

\section{Discussion and Analysis}
Section \ref{subsec:cross-vs-mutate-operators} compared different crossover and mutation operators. In these experiments \textit{OrderCrossover} was consistently the best crossover operator out of the three. This matches expectations because OX is designed to preserve the order of chromosome elements. In problems such as the TSP the order is perhaps the most import part of the solution rather than just the producing valid tours. I would suggest that OX better encodes heuristics about the problem domain compared to PMX.

Comparing the mutation operators shows that \textit{InversionMutation} appears to produce the best results and \textit{SwapCityMutation} produces the worst. Logically the \textit{SwapCityMutation} should be more destructive than the displacement and inversion operations which keep whole portions of the chromosome in order. As order is key to solving the TSP these mutations must be able to push the GA to explore the search space effectively without being overly destructive. It's worth noting that the \textit{InsertionMutation} achieves results that are very similar to the \textit{DisplacementMutation} and much better than just swapping cities. This was slightly unexpected. It could just be that this operator performed better because it's slightly less destructive to the order or it could just be a statistical fluke.

The crossover and mutation rate results (section \ref{subsec:crossover-mutation-parameters}) clearly shows a tendency towards larger crossover and mutation rates. Looking at table \ref{table:cross-vs-mutate-params} it can be seen that the results definitively become better towards the bottom right. What is also interesting in this result is that the effect of changing the mutation rate has a larger effect on the results than the crossover rate. Comparing this with figure \ref{fig:cross-vs-mutate-params} shows that increasing the mutation rate increases the chance of the GA finding a better solution. 

The effect of the crossover rate appears to be more subtle. Looking at figure \ref{fig:cross-vs-mutate-params} shows that a very high crossover rate ($0.9$) causes the GA to plateau earlier. I.e. a high crossover rate causes the GA to converge faster but not necessarily to the best solution. However, when combined with a high mutation rate, the convergence becomes more consistent, although not identical.

Testing the selection operator shows some interesting insights. Looking at table \ref{table:selection-vs-pop-size} and figure \ref{fig:tournament-selection-convergence} a clear relationship can be seen between the population size used and the tournament size. Best results were achieved with both a large population and tournament size. Another interesting point is that like the relationship between the mutation and crossover parameters the difference between the values becomes smaller as the parameters increase. This suggests two things 1)  there is an optimum setting between the two parameters and 2) that increasing them indefinitely will probably not significantly improve GA performance. Also, in figure \ref{fig:tournament-selection-convergence} it can be seen that a very small tournament size ($3$) stifles the diversity and makes it harder for the algorithm to make progress.

The results in table \ref{fig:knn-convergence} show the a direct relationship where the fitness of the population decreases in proportion to the number of KNN based chromosomes used. This is backed up by figure \ref{fig:knn-convergence} which shows a difference in the initial rate of convergence between random population generation and all settings of KNN population generation. As predicted this difference was more prominent with a greater number of cities hence why $100$ cities where used.

\section{Conclusions and Future Work}
In conclusion, a few interesting global trends have been identified in applying GAs to the TSP. One trend appears to be that the best choice of crossover and mutation operators are ones which maximally preserve the order of sub-tours. This might prevent the algorithm from loosing good portions of candidate solutions and convergence on local minima.

Secondly, it seems that high crossover and mutation rates are preferred. In particular the mutation rate seems to make the highest impact on the rate at which a GA stabilises on a particular solution. The crossover rate also has an impact, but it's effect is less prominent.

TS appeared was more stable than RWS. With tournament selection a larger population size and tournament size increase performance. However, the tournament size and in particular the population size significantly impacts the time complexity. Combined with the fact that increasing these parameters doesn't linearly increase performance  confirms that a balance is required between selection parameters and computation time.

The use of KNN population generation can provide a useful increase to the performance of an algorithm. KNN population generation can lead to faster convergence and better solutions by starting with better chromosomes. But a balance must be struck to prevent decreasing diversity.

GAs appear to be well suited to the problem of solving the TSP. With this system good solutions could be found for 30 to 50 cities, but quality deteriorated beyond this. However, the requirement that the genes must be unique and that ordering is important is a handicap for GAs. It would be interesting to investigate the effect of ant colony optimisation on the TSP as this approach more naturally lends itself to path finding in comparison to a GA.

Finally, there are several areas which have not been fully explored here, but which could be explored in future work. This includes modifying the implementation to use multiprocessing. GAs naturally lead themselves to parallelisation and such an implementation would make it easier to run larger parameter sets. Another idea would be to explore modification of the population and parameters over time. This could use approaches such as simulated annealing to control crossover \& mutation parameters or to reintroduce diversity using approaches such as random offspring generation \cite{rocha1999preventing} or social disasters \cite{kureichick1996some}.


% needed in second column of first page if using \IEEEpubid
%\IEEEpubidadjcol

% An example of a floating figure using the graphicx package.
% Note that \label must occur AFTER (or within) \caption.
% For figures, \caption should occur after the \includegraphics.
% Note that IEEEtran v1.7 and later has special internal code that
% is designed to preserve the operation of \label within \caption
% even when the captionsoff option is in effect. However, because
% of issues like this, it may be the safest practice to put all your
% \label just after \caption rather than within \caption{}.
%
% Reminder: the "draftcls" or "draftclsnofoot", not "draft", class
% option should be used if it is desired that the figures are to be
% displayed while in draft mode.
%
%\begin{figure}[!t]
%\centering
%\includegraphics[width=2.5in]{myfigure}
% where an .eps filename suffix will be assumed under latex, 
% and a .pdf suffix will be assumed for pdflatex; or what has been declared
% via \DeclareGraphicsExtensions.
%\caption{Simulation Results}
%\label{fig_sim}
%\end{figure}

% Note that IEEE typically puts floats only at the top, even when this
% results in a large percentage of a column being occupied by floats.


% An example of a double column floating figure using two subfigures.
% (The subfig.sty package must be loaded for this to work.)
% The subfigure \label commands are set within each subfloat command, the
% \label for the overall figure must come after \caption.
% \hfil must be used as a separator to get equal spacing.
% The subfigure.sty package works much the same way, except \subfigure is
% used instead of \subfloat.
%
%\begin{figure*}[!t]
%\centerline{\subfloat[Case I]\includegraphics[width=2.5in]{subfigcase1}%
%\label{fig_first_case}}
%\hfil
%\subfloat[Case II]{\includegraphics[width=2.5in]{subfigcase2}%
%\label{fig_second_case}}}
%\caption{Simulation results}
%\label{fig_sim}
%\end{figure*}
%
% Note that often IEEE papers with subfigures do not employ subfigure
% captions (using the optional argument to \subfloat), but instead will
% reference/describe all of them (a), (b), etc., within the main caption.


% An example of a floating table. Note that, for IEEE style tables, the 
% \caption command should come BEFORE the table. Table text will default to
% \footnotesize as IEEE normally uses this smaller font for tables.
% The \label must come after \caption as always.
%
%\begin{table}[!t]
%% increase table row spacing, adjust to taste
%\renewcommand{\arraystretch}{1.3}
% if using array.sty, it might be a good idea to tweak the value of
% \extrarowheight as needed to properly center the text within the cells
%\caption{An Example of a Table}
%\label{table_example}
%\centering
%% Some packages, such as MDW tools, offer better commands for making tables
%% than the plain LaTeX2e tabular which is used here.
%\begin{tabular}{|c||c|}
%\hline
%One & Two\\
%\hline
%Three & Four\\
%\hline
%\end{tabular}
%\end{table}


% Note that IEEE does not put floats in the very first column - or typically
% anywhere on the first page for that matter. Also, in-text middle ("here")
% positioning is not used. Most IEEE journals use top floats exclusively.
% Note that, LaTeX2e, unlike IEEE journals, places footnotes above bottom
% floats. This can be corrected via the \fnbelowfloat command of the
% stfloats package.







% if have a single appendix:
%\appendix[Proof of the Zonklar Equations]
% or
%\appendix  % for no appendix heading
% do not use \section anymore after \appendix, only \section*
% is possibly needed

% use appendices with more than one appendix
% then use \section to start each appendix
% you must declare a \section before using any
% \subsection or using \label (\appendices by itself
% starts a section numbered zero.)
%


\appendices
\clearpage

\section{Installation of Program}
\label{appendix:installation}
Installation is easiest using the \textit{pip} package manager. Python version $2.7.9+$ automatically ships with \textit{pip}. If you're using an older version of Python you can find the installation instructions at http://pip.readthedocs.org/en/stable/installing/.

To install the program locally \textit{cd} into the top level directory of the project and run the following:

\begin{lstlisting}[language=Bash]
pip install -e .	
\end{lstlisting}


All of the project dependancies should be installed automatically. If for some reason they are not, the full list of dependancies are:

\begin{itemize}
	\item numpy
	\item scipy
	\item pandas
	\item scikit-learn
	\item matplotlib
	\item click
\end{itemize}

\section{Command Line Interface}
\label{appendix:cli}
There are three major commands: \textit{generate}, \textit{solve} and \textit{tune}. 

\subsection{Generating Datasets}
The generate command will create a uniformly random TSP dataset and output it to the specified CSV file.

\begin{lstlisting}[language=Bash]
tspsolver generate dataset.csv	
\end{lstlisting}

Optionally, it can also take a parameter specifying the number of cities to generate (default is 10):

\begin{lstlisting}[language=Bash]
tspsovler generate -n 30 dataset.csv	
\end{lstlisting}


\subsection{Solving TSP problems}
The \textit{solve} command can be used to run the genetic algorithm and produce solutions to TSP problems. The command takes a JSON parameter file as an argument. Optionally you can provide a dataset file (such as produced by the \textit{generate} command) or specify a random number of cities to generate. When the command terminates, two plots are produced. One shows the min, mean, and max fitness across all generations, the second shows the best solution found over all datasets.

Example Parameter File:
\begin{lstlisting}
{
    "num_epochs": 1000,
    "num_elites": 2,
    "generator": "SimplePopulationGenerator",
    "generator_population_size": 20,
    "selector": "RouletteWheelSelection",
    "selector_tournament_size": 5,
    "crossover": "OrderCrossover",
    "crossover_pcross": 0.9,
    "mutator": "InversionMutation",
    "mutator_pmutate": 0.1
}
\end{lstlisting}

Solving a TSP problem with a randomly generated TSP dataset:

\begin{lstlisting}[language=Bash]
tspsolver solve -n 20  params.json
\end{lstlisting}

Solving a TSP problem with an exisitng dataset:
\begin{lstlisting}[language=Bash]
tspsolver solve -f dataset.csv params.json
\end{lstlisting}

\subsection{Tuning Parameters}
The final command can be used to run a range of parameter configurations over a number of different datasets. This can be useful to examine the effects of different parameter datasets. Each configuration is run on $n$ different randomly generated datasets and the median result is taken to represent the whole. The results for all datasets are saved to a CSV file. The configuration that produced the best results is also saved to a JSON file.

This command takes a special parameter file that specifies ranges of parameters. An example is shown below:

\begin{lstlisting}
{
    "num_epochs": [1000],
    "num_elites": [0],
    "generator": ["SimplePopulationGenerator"],
    "generator_population_size": [20],
    "selector": ["TournamentSelection"],
    "selector_tournament_size": [5],
    "crossover": ["OrderCrossover"],
    "crossover_pcross": [0.6, 0.7, 0.8, 0.9],
    "mutator": ["InversionMutation"],
    "mutator_pmutate": [0.01, 0.05, 0.1, 0.2]
}
\end{lstlisting}

This will run the genetic algorithm with a varying range of crossover and mutation probabilities. An example of running the tuning command is as follows:

\begin{lstlisting}[language=Bash]
tspsolver tune -d 5 -n 50 tuning_params.json results.csv best.json
\end{lstlisting}

In the above command the \textit{-d} command specifies the number of datasets to generate for each parameter configuration. The \textit{-n} flag specifies the number of random generated points to use for each dataset. \textit{tuning\_params.json} is the special parameter file with ranges. \textit{results.csv} is the CSV file created with all parameter results. \textit{best.json} is the generated parameter file containing the parameters that produced the best run.

\clearpage
\onecolumn
\section{Parameters}
\label{appendix:parameters}
This appendix provides an overview of the different parameters that can be used with the system with example values and a description of what the parameter controls.


\begin{table}[ht]
    \begin{tabular}{l | l | p{11cm} }
    Parameter                   & Example Value              & Description                                                                                                                               \\ \hline
    num\_epochs                 & 1000                       & The number of epochs to run the genetic algorithm for.                                                                                    \\ 
    num\_elites                 & 2                          & The number of elites to carry over the the next generation.                                                                               \\
    generator                   & SimplePopulationGeneration & The type of generator to use to initialise the population. This can either be \textit{SimplePopulationGeneration} or \textit{KNNPopulationGenerator}.       \\
    generator\_population\_size & 20                         & Specifies the population size of the genetic algorithm                                                                                    \\
    generator\_random\_proportion & 0.5						 & When using KNN population generation this controls the ratio of individuals selected using KNN relative to the number of random chromosomes. Must be in the range $0 \leq x \leq 1$ \\ 
    selector                    & TournamentSelection        & Specifies which type of selection to use. This can either be \textit{TournamentSelection} or \textit{RouletteWheelSelection}                                \\
    selector\_tournament\_size  & 5                          & When using tournament selection this controls the size of the tournament used.                                                            \\
    crossover                   & OrderCrossover             & Specifies the crossover operator to use. This can be one of: \textit{OrderCrossover}, \textit{OnePointPMX}, or \textit{TwoPointPMX}.                                 \\
    crossover\_pcross           & 0.9                        & This controls the probability that crossover will occur.  Must be in the range $0 \leq x \leq 1$                                                \\
    mutator                     & InversionMutation          & Specifies the mutation operator to use. This can be one of: \textit{SwapCityMutation}, \textit{InversionMutation}, \textit{InsertionMutation}, \textit{DisplacementMutation}. \\
    mutator\_pmutate            & 0.1                        & This controls the probability that mutation will occur. Must be in the range $0 \leq x \leq 1$                                                  \\
    \end{tabular}
\end{table}

\clearpage
\twocolumn

%Some text for the appendix.

% use section* for acknowledgement
%\section*{Acknowledgment}
%
%
%The authors would like to thank...


% Can use something like this to put references on a page
% by themselves when using endfloat and the captionsoff option.
\ifCLASSOPTIONcaptionsoff
  \newpage
\fi



% trigger a \newpage just before the given reference
% number - used to balance the columns on the last page
% adjust value as needed - may need to be readjusted if
% the document is modified later
%\IEEEtriggeratref{8}
% The "triggered" command can be changed if desired:
%\IEEEtriggercmd{\enlargethispage{-5in}}

% references section

% can use a bibliography generated by BibTeX as a .bbl file
% BibTeX documentation can be easily obtained at:
% http://www.ctan.org/tex-archive/biblio/bibtex/contrib/doc/
% The IEEEtran BibTeX style support page is at:
% http://www.michaelshell.org/tex/ieeetran/bibtex/
%\bibliographystyle{IEEEtran}
% argument is your BibTeX string definitions and bibliography database(s)
%\bibliography{IEEEabrv,../bib/paper}
%
% <OR> manually copy in the resultant .bbl file
% set second argument of \begin to the number of references
% (used to reserve space for the reference number labels box)
\bibliographystyle{plain}
\bibliography{references}
%\begin{thebibliography}{1}
%
%\bibitem{IEEEhowto:kopka}
%H.~Kopka and P.~W. Daly, \emph{A Guide to \LaTeX}, 3rd~ed.\hskip 1em plus
%  0.5em minus 0.4em\relax Harlow, England: Addison-Wesley, 1999.
%
%\end{thebibliography}

% biography section
% 
% If you have an EPS/PDF photo (graphicx package needed) extra braces are
% needed around the contents of the optional argument to biography to prevent
% the LaTeX parser from getting confused when it sees the complicated
% \includegraphics command within an optional argument. (You could create
% your own custom macro containing the \includegraphics command to make things
% simpler here.)
%\begin{biography}[{\includegraphics[width=1in,height=1.25in,clip,keepaspectratio]{mshell}}]{Michael Shell}
% or if you just want to reserve a space for a photo:

%\begin{IEEEbiography}[{\includegraphics[width=1in,height=1.25in,clip,keepaspectratio]{picture}}]{John Doe}
%\blindtext
%\end{IEEEbiography}

% You can push biographies down or up by placing
% a \vfill before or after them. The appropriate
% use of \vfill depends on what kind of text is
% on the last page and whether or not the columns
% are being equalized.

%\vfill

% Can be used to pull up biographies so that the bottom of the last one
% is flush with the other column.
%\enlargethispage{-5in}



% that's all folks
\end{document}


